File: lib/models/notification_preferences.dart
import 'package:json_annotation/json_annotation.dart';

part 'notification_preferences.g.dart';

/// Notification preferences model for managing user notification categories.
///
/// **Requirements Coverage:**
/// - Requirement 38.2: Preferences with categories (showtimeReminders, promotions, newMovies, bookingUpdates)
/// - Requirement 38.6: Toggle switches for each category with descriptive labels
///
/// Each boolean field represents a notification category:
/// - `showtimeReminders`: 2-hour reminder before showtime
/// - `promotions`: Marketing and promotional notifications
/// - `newMovies`: New movie release announcements
/// - `bookingUpdates`: Booking confirmations and status changes
///
/// Usage:
/// ```dart
/// // From JSON
/// final prefs = NotificationPreferences.fromJson(jsonMap);
///
/// // Update preference
/// final updated = prefs.copyWith(promotions: false);
/// ```
@JsonSerializable()
class NotificationPreferences {
  /// Enable showtime reminder notifications (2 hours before showtime)
  final bool showtimeReminders;

  /// Enable promotional and marketing notifications
  final bool promotions;

  /// Enable new movie release notifications
  final bool newMovies;

  /// Enable booking update notifications (confirmations, changes)
  final bool bookingUpdates;

  const NotificationPreferences({
    required this.showtimeReminders,
    required this.promotions,
    required this.newMovies,
    required this.bookingUpdates,
  });

  /// Creates an instance from JSON map
  factory NotificationPreferences.fromJson(Map<String, dynamic> json) =>
      _$NotificationPreferencesFromJson(json);

  /// Converts instance to JSON map
  Map<String, dynamic> toJson() => _$NotificationPreferencesToJson(this);

  /// Creates a copy with optional field overrides
  NotificationPreferences copyWith({
    bool? showtimeReminders,
    bool? promotions,
    bool? newMovies,
    bool? bookingUpdates,
  }) {
    return NotificationPreferences(
      showtimeReminders: showtimeReminders ?? this.showtimeReminders,
      promotions: promotions ?? this.promotions,
      newMovies: newMovies ?? this.newMovies,
      bookingUpdates: bookingUpdates ?? this.bookingUpdates,
    );
  }

  /// Default preferences (all enabled)
  static const NotificationPreferences defaults = NotificationPreferences(
    showtimeReminders: true,
    promotions: true,
    newMovies: true,
    bookingUpdates: true,
  );

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is NotificationPreferences &&
        other.showtimeReminders == showtimeReminders &&
        other.promotions == promotions &&
        other.newMovies == newMovies &&
        other.bookingUpdates == bookingUpdates;
  }

  @override
  int get hashCode {
    return showtimeReminders.hashCode ^
        promotions.hashCode ^
        newMovies.hashCode ^
        bookingUpdates.hashCode;
  }
}

File: backend/src/main/java/com/cineluxe/entity/NotificationPreferences.java
package com.cineluxe.entity;

import jakarta.persistence.*;
import java.time.LocalDateTime;

/**
 * JPA Entity representing user notification preferences.
 * 
 * Requirements Coverage:
 * - Requirement 38.2: Preferences with categories (showtimeReminders, promotions, newMovies, bookingUpdates)
 * - Each category is a boolean flag indicating whether the user wants to receive that type of notification
 * 
 * One-to-one relationship with User entity (userId is unique).
 */
@Entity
@Table(name = "notification_preferences")
public class NotificationPreferences {
    
    /**
     * User ID - primary key ensures one preference record per user
     */
    @Id
    @Column(name = "user_id", nullable = false)
    private String userId;
    
    /**
     * Enable showtime reminder notifications (2 hours before showtime)
     */
    @Column(name = "showtime_reminders", nullable = false)
    private Boolean showtimeReminders = true;
    
    /**
     * Enable promotional and marketing notifications
     */
    @Column(name = "promotions", nullable = false)
    private Boolean promotions = true;
    
    /**
     * Enable new movie release notifications
     */
    @Column(name = "new_movies", nullable = false)
    private Boolean newMovies = true;
    
    /**
     * Enable booking update notifications (confirmations, changes)
     */
    @Column(name = "booking_updates", nullable = false)
    private Boolean bookingUpdates = true;
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }
    
    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
    
    // Constructors
    
    public NotificationPreferences() {
    }
    
    public NotificationPreferences(String userId) {
        this.userId = userId;
        // Default all preferences to true
        this.showtimeReminders = true;
        this.promotions = true;
        this.newMovies = true;
        this.bookingUpdates = true;
    }
    
    // Getters and Setters
    
    public String getUserId() {
        return userId;
    }
    
    public void setUserId(String userId) {
        this.userId = userId;
    }
    
    public Boolean getShowtimeReminders() {
        return showtimeReminders;
    }
    
    public void setShowtimeReminders(Boolean showtimeReminders) {
        this.showtimeReminders = showtimeReminders;
    }
    
    public Boolean getPromotions() {
        return promotions;
    }
    
    public void setPromotions(Boolean promotions) {
        this.promotions = promotions;
    }
    
    public Boolean getNewMovies() {
        return newMovies;
    }
    
    public void setNewMovies(Boolean newMovies) {
        this.newMovies = newMovies;
    }
    
    public Boolean getBookingUpdates() {
        return bookingUpdates;
    }
    
    public void setBookingUpdates(Boolean bookingUpdates) {
        this.bookingUpdates = bookingUpdates;
    }
    
    public LocalDateTime getCreatedAt() {
        return createdAt;
    }
    
    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }
    
    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }
    
    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}
